src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -603,6 +655,10 @@ void CallGenerator::do_late_inline_helper() {
     C->remove_macro_node(call);
   }
 
+  if (C->eliminate_boxing()) {
+    replace_box_to_scalar(call, callprojs.resproj);
```

Comment:
Another confusing aspect: what happens with `call` node after `replace_box_to_scalar` call is over?

My understanding is you rely on `replace_box_to_scalar` to eliminate all the uses by safepoints and then let  the following code to clean it up:
```
  bool result_not_used = (callprojs.resproj == NULL || callprojs.resproj->outcnt() == 0);
  if (is_pure_call() && result_not_used) {
    // The call is marked as pure (no important side effects), but result isn't used.
    // It's safe to remove the call.
    GraphKit kit(call->jvms());
    kit.replace_call(call, C->top(), true);
```

The relation between `is_pure_call()` and `C->eliminate_boxing()`/`call->as_CallStaticJava()->is_boxing_method()` is not clear.

I suggest to refactor `is_pure_call()` and add a special case for `is_boxing_method()` to perform scalarization when there are only debug usages. 

Also, add an assert in/after `replace_box_to_scalar` ensuring there are no users left, so the call can go away.

Reply:
Regarding `C->eliminate_boxing()`: why do you think it is needed? 

I'm asking because `is_pure_call() && result_not_used` doesn't have that check.

My understanding is the code encounters `is_boxing_method()` there only when `Compile::should_delay_boxing_inlining() == true` which requires both `EliminateAutoBox` and `AggressiveUnboxing`

Reply:
I will refactor these codes and add `is_boxing_call()` method. Thank you very much for your review.

Reply:
I just hit bug where VectorReinterpretTest.java test run with -XX:-DoEscapeAnalysis but Vector API generates SafePointScalarObjectNode regardless that:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vector.cpp#L249

But during deoptimization we only restore allocations when DoEscapeAnalysis is true:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/deoptimization.cpp#L377

(currently condition is always true because jvmci_enabled == true since JVMCI included into build but without JVMCI it will fail as I found).

It seems with this change you need to add EliminateAutoBox flag check to deoptimization condition.
Or guard your optimization with C->do_escape_analysis().

@iwanowww  Do you think we should add checks for EnableVectorAggressiveReboxing and  EliminateAutoBox into deoptimization code or guard Vector API code and this optimization with  C->do_escape_analysis()?

Reply:
> I just hit bug where VectorReinterpretTest.java test run with -XX:-DoEscapeAnalysis but Vector API generates SafePointScalarObjectNode regardless that:
> https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vector.cpp#L249
> 
> But during deoptimization we only restore allocations when DoEscapeAnalysis is true:
> https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/deoptimization.cpp#L377
> 
> (currently condition is always true because jvmci_enabled == true since JVMCI included into build but without JVMCI it will fail as I found).
> 
> It seems with this change you need to add EliminateAutoBox flag check to deoptimization condition.
> Or guard your optimization with C->do_escape_analysis().
> 
> @iwanowww Do you think we should add checks for EnableVectorAggressiveReboxing and EliminateAutoBox into deoptimization code or guard Vector API code and this optimization with C->do_escape_analysis()?

The issue was fixed: 8263125. https://git.openjdk.java.net/jdk/pull/2924